upstairs/src/client.rs

```diff
@@ -2916,108 +2916,51 @@ pub(crate) fn validate_encrypted_read_response(
-            &block_encryption_ctx.tag[..],
+    };
+
+    let Some(block_encryption_ctx) = &context.encryption_context else {
+        // this block context is missing an encryption context!
+        // XXX should this be an error instead?
+        return Err(CrucibleError::DecryptionError);
+    };
+
+    // Validate integrity hash before decryption
+    let computed_hash = integrity_hash(&[
+        &block_encryption_ctx.nonce[..],
+        &block_encryption_ctx.tag[..],
+        data,
+    ]);
+
+    if computed_hash == context.hash {
+        // Now that the integrity hash was verified, attempt
+        // decryption.
+        //
+        // Note: decrypt_in_place does not overwrite the buffer if
+        // it fails, otherwise we would need to copy here. There's a
+        // unit test to validate this behaviour.
+        use aes_gcm_siv::{Nonce, Tag};
+        let decryption_result = encryption_context.decrypt_in_place(
             data,
-        ]);
-
-        if computed_hash == ctx.hash {
-            valid_hash = Some(ctx.hash);
-
-            // Now that the integrity hash was verified, attempt
-            // decryption.
-            //
-            // Note: decrypt_in_place does not overwrite the buffer if
-            // it fails, otherwise we would need to copy here. There's a
-            // unit test to validate this behaviour.
-            use aes_gcm_siv::{Nonce, Tag};
-            let decryption_result = encryption_context.decrypt_in_place(
-                data,
-                Nonce::from_slice(&block_encryption_ctx.nonce[..]),
-                Tag::from_slice(&block_encryption_ctx.tag[..]),
-            );
-
-            if decryption_result.is_ok() {
-                successful_decryption = true;
-                break;
-            } else {
-                // Only one hash + nonce + tag combination will match the
-                // data that is returned. Due to the fact that nonces are
-                // random for each write, even if the Guest wrote the
-                // same data block 100 times, only one index will be
-                // valid.
-                //
-                // If the computed integrity hash matched but decryption
-                // failed, continue to the next contexts. the current
-                // hashing algorithm (xxHash) is not a cryptographic hash
-                // and is only u64, so collisions are not impossible.
-                warn!(
-                    log,
-                    "Decryption failed even though integrity hash matched!"
-                );
-            }
-        }
-    }
-
-    if let Some(valid_hash) = valid_hash {
-        if !successful_decryption {
+            Nonce::from_slice(&block_encryption_ctx.nonce[..]),
+            Tag::from_slice(&block_encryption_ctx.tag[..]),
+        );
+        if decryption_result.is_ok() {
+            Ok(Some(context.hash))
+        } else {
             // No encryption context combination decrypted this block, but
```

Comment:
How do you feel about ignoring the hash and doing decryption only (for encrypted Downstairs)?

"Decryption succeeds + hash fails" doesn't seem like a case that we care about protecting against, since we're using authenticated encryption, and it would be free ~real estate~ cycles

Reply:
It seems like if the decryption works, then it is all good. 
I'm not even sure what it would mean if the hash fails on a good decryption?  Data corruption in the on disk (or in memory) hash?

What if we only check the integrity hash if the decryption fails?  In that case we are in the error path
anyway so performance is not as criticial, and it might give us a bit more info on the error (like, knowing we
have a bad key vs. something else?  maybe?)

This brings up another question, why do we have the integrity hash at all any longer?  Do we need it?
It does allow for a downstairs to check a block without having to know the key, but I don't believe we
ever use that anywhere.

Reply:
I've changed the logic for encrypted block to pass if decryption succeeds.  To compare read results between different Downstairs, we now store an
```rust
pub(crate) enum Validation {
    /// The block has no hash / context and is empty
    Empty,
    /// For an unencrypted block, the result is the hash
    Unencrypted(u64),
    /// For an encrypted block, the result is the tag + nonce
    Encrypted(crucible_protocol::EncryptionContext),
}
```
(instead of the prior `Option<u64>`, which is just the hash)

This is done in 56f7529195b6

Reply:
> This brings up another question, why do we have the integrity hash at all any longer? Do we need it?
It does allow for a downstairs to check a block without having to know the key, but I don't believe we
ever use that anywhere.

It does allow for checking for on-the-wire bit flips by either the Upstairs or Downstairs. We could work this into some sort of re-transmission scheme?

Reply:
"Integrity hash" is ambiguous, since there are actually two hashes (and three different use cases):

- The `on_disk_hash` lets the Downstairs decide which of the two context slots is valid, if there's a dirty shutdown
- To detect in-transit corruption from the Upstairs → Downstairs, we use the per-block hash (included in the `Message`)
- To detect in-transit corruption in the Downstairs → Upstairs path, we use
  - The per-block hash, for unencrypted blocks
  - The encryption context, for encrypted blocks; the hash is also transmitted, but is now only checked if decryption fails, to distinguish which error message to return

(see RFD 490 for a modest proposal to further simplify the system)